evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -8,28 +14,146 @@
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
     decode_hex,
 )
+from evm.utils.keccak import (
+    keccak,
+)
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     DEPOSIT_SIZE,
     GASPRICE,
+    PERIOD_LENGTH,
     TX_GAS,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+class FilterNotFound(Exception):
+    pass
+
+
 class VMC(Contract):
 
     logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
 
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    collation_added_topic = "0x" + keccak(b"CollationAdded(int128,bytes4096,bool,int128)").hex()
+    # shard_id -> list
+    # older <---------------> newer
+    new_collation_added_logs = {}
+    # shard_id -> list
+    # newer <---------------> older
+    unchecked_collation_added_logs = {}
+    # shard_id -> filter
+    collation_added_filter = {}
+    # shard_id -> score
+    current_checking_score = {}
+
     def __init__(self, *args, default_privkey, **kwargs):
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         super().__init__(*args, **kwargs)
 
+    def setup_collation_added_filter(self, shard_id):
+        shard_id_topic = "0x" + shard_id.to_bytes(32, byteorder='big').hex()
+        self.collation_added_filter[shard_id] = self.web3.eth.filter({
+            'address': self.address,
+            'topics': [
+                self.collation_added_topic,
+                shard_id_topic,
+            ],
+        })
+        self.new_collation_added_logs[shard_id] = []
+        self.unchecked_collation_added_logs[shard_id] = []
+        self.current_checking_score[shard_id] = None
+
+    @to_dict
+    def parse_collation_added_data(self, data):
+        score = big_endian_to_int(data[-32:])
+        is_new_head = bool(big_endian_to_int(data[-64:-32]))
+        header_bytes = data[:-64]
+        # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+        sedes = rlp.sedes.List([
+            rlp.sedes.big_endian_int,
+            rlp.sedes.big_endian_int,
+            hash32,
+            hash32,
+            hash32,
+            address,
+            hash32,
+            hash32,
+            rlp.sedes.big_endian_int,
+            rlp.sedes.binary,
+        ])
+        header_values = rlp.decode(header_bytes, sedes=sedes)
+        yield 'header', header_values
+        yield 'is_new_head', is_new_head
+        yield 'score', score
+
+    @to_tuple
+    def _get_new_logs(self, shard_id):
+        # use `get_new_entries` over `get_all_entries`
+        #   1. Prevent from the increasing size of logs
+        #   2. Leave the efforts maintaining `new_logs` in RPC servers
```

Comment:
I have a pattern I've been wanting to try out that this might be a good candidate for.  It's meant to emulate `get_new_entries` without any of the complication of installing the filter and tracking the `filter_id`.  It looks roughly like this.

```
from cytoolz import (
    merge,
    sliding_window,
)


def log_filter(web3, filter_params, page_size=10):
    # TODO: handle pending/latest/earliest
    from_block = filter_params['fromBlock']
    to_block = filter_params['toBlock']

    # TODO: handle dynamic block ranges to enforce reasonable request sizes.
    for block_number in range(from_block, to_block + 1):
        params = merge(filter_params, {'fromBlock': block_number, 'toBlock': block_number})
        logs = web3.eth.getLogs(params)

        for page in sliding_window(page_size, logs):
            yield page
```

Issue in web3.py here: https://github.com/ethereum/web3.py/issues/551

I suspect you could implement a specialized version of this that would be more robust than relying on filters on the server, eliminating the need to handle things like re-installing the filter if it gets lost.

Reply:
No problem.
Thank you for letting me know of this. I didn't consider the situation that filters get lost in servers.
I will do it. :)

Reply:
Since I just found `getLogs` has not been implemented in eth_tester, I copied and pasted some code from `eth_tester.create_log_filter` and `eth_tester._process_block_logs` to a new function [`eth_tester.get_logs`](https://gist.github.com/mhchia/45e7fd2ada48918eb2bbb0c6a549017a), and made `web3.py` calls it. Now it seems to work fine.(still need testing).
One question is, for now I plan to save `block_number` every time I get new logs with `eth_getLogs`, to identify the latest block which our new logs have been synced to. Should I handle the situation when forks happen? If we take forks in mainchain into consideration, it seems to be more complicated, and we need to store more things.
Thank you for your wisdom:)

Reply:
@pipermerriam I'm sorry that after thinking for a while, I still have some questions about the code.
1. We should specify the `fromBlock` and `toBlock` to ensure the logs we are going to get are in the blocks range.
2. Why not just
```
    for block_number in range(from_block, to_block + 1):
        params = merge(filter_params, {'fromBlock': block_number, 'toBlock': block_number})
        logs = web3.eth.getLogs(params)
        for page in logs:
            yield page
```
instead of 
```
    for block_number in range(from_block, to_block + 1):
        params = merge(filter_params, {'fromBlock': block_number, 'toBlock': block_number})
        logs = web3.eth.getLogs(params)
        for page in sliding_window(page_size, logs):
            yield page
```
? What is purpose to use the `sliding_window`?
Thank you so much!

Reply:
I have no idea why I put `sliding_window` in there.  I think I got my wires crossed while hammering out that example code.  Will try to get you answers to other questions asap but I'm out of time today.

Reply:
Instead of sliding window, it should be [`toolz.partition_all`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition_all)

Reply:
hmm that makes sense.
Thank you so much.
Please take your time. It is no hurry at all!

Reply:
> I have one question is, for now I plan to save `block_number` every time I get new logs with `eth_getLogs`, to identify the latest block which our new logs have been synced to. Should I handle the situation when forks happen? If we take forks in mainchain into consideration, it seems to be more complicated, and we need to store more things.

Here is some untested code that won't actually work but should demonstrate the idea:

```python
def get_previous_headers(w3, history_size=256):
    block = w3.eth.getBlock('latest')
    recent_hashes = [block['hash']]

    # initialize the list of recent hashes
    for _ in range(history_size - 1):
        block = w3.eth.getBlock(block['parentHash'])
        recent_hashes.append(block['hash'])

        # break the loop if we hit the genesis block.
        if block['number'] == 0:
            break

    recent_hashes = tuple(reversed(_recent_hashes))
    return recent_hashes


def check_chain_head(w3, known_previous_headers):
    block = w3.eth.getBlock('latest')

    new_blocks = []

    for _ in range(history_size):
        if block['hash'] in recent_hashes:
            break
        new_block_hashs.append(block['hash'])
        block = w3.eth.getBlock(block['parentHash'])
    else:
        raise Exception('No common ancestor found for block: {0}'.format(block['hash']))

    first_common_ancestor_idx = recent_hashes.index(block['hash'])

    unchanged_hashes = recent_hashes[:first_common_ancestor_idx]
    revoked_hashes = recent_hashes[first_common_ancestor_idx:]

    recent_hashes.extend(new_block_hashs)
    # this won't actually work because of variable scoping
    recent_hashs = recent_hashs[-1 * history_size:]
    return revoked_hashes, new_block_hashs


```

The idea is to build a utility that will track the most recent hashes in the head of the chain in such a way that it can provide you with two values.

- the block hashes that were revoked from the head of the chain due to re-org
- the new block hashes that we haven't yet seen

You can use the revoked hashes to roll back any logs that you processed which are no longer valid (assuming you need to do that).

You can use the new hashes to define the `fromBlock` and `toBlock` for fetching any new logs that you haven't yet seen.

Questions?  Does this make sense?  The code above could probably more easily be written as a class.

Reply:
It's a beautiful and simple idea! In this way we can focus on the `revoked_changes` and `unseen_and_canonical_changes`, and no need to worry about the other re-orgs between them. I think I can soon implement the utility based on your code.
Still one question: Why is `history_size` set to `256`? Should it be larger or smaller?
I really appreciate your help and patience!

Reply:
`256` was a totally arbitrary number.  Should probably be configurable.